distributed/protocol/serialize.py

```diff
@@ -450,30 +451,74 @@ def serialize_and_split(
     return header, out_frames
 
 
-def merge_and_deserialize(header, frames, deserializers=None):
+def merge_and_deserialize(header, frames, deserializers=None, memoryview_offset=0):
     """Merge and deserialize frames
 
-    This function is a drop-in replacement of `deserialize()` that merges
-    frames that were split by `serialize_and_split()`
+    This function is a replacement for `deserialize()` that merges
+    frames that were split by `serialize_and_split()`.
+
+    When ``frames`` contains memoryviews that share an underlying buffer,
+    ``memoryview_offset`` must be the index into that underlying buffer
+    where the ``frames`` starts (in bytes, not frame counts).
 
     See Also
     --------
     deserialize
     serialize_and_split
+    merge_subframes
     """
     merged_frames = []
     if "split-num-sub-frames" not in header:
         merged_frames = frames
     else:
+        frame_byte_offsets = list(itertools.accumulate(map(len, frames)))
         for n, offset in zip(header["split-num-sub-frames"], header["split-offsets"]):
             if n == 1:
-                merged_frames.append(frames[offset])
+                merged = frames[offset]
             else:
-                merged_frames.append(bytearray().join(frames[offset : offset + n]))
+                subframes = frames[offset : offset + n]
+                merged = merge_subframes(
+                    subframes,
+                    memoryview_offset=memoryview_offset
+                    + (frame_byte_offsets[offset - 1] if offset else 0),
+                )
+            merged_frames.append(merged)
+            memoryview_offset += len(merged)
 
     return deserialize(header, merged_frames, deserializers=deserializers)
 
 
+def merge_subframes(
+    subframes: "list[memoryview | bytearray | bytes]", memoryview_offset: int = 0
+) -> "memoryview | bytearray":
```

Comment:
You can remove the quote marks from the type annotations if your module starts with ``from __future__ import annotations`` (note: won't work with scheduler.py due to cythonization)

Reply:
Good point, I was just lazy :)